src/perf/best-practices.md

```diff
@@ -0,0 +1,425 @@
+performance. These best practice recommendations will help you
+write the most performant Flutter app possible.
+
+{{site.alert.note}}
+  If you are writing web apps in Flutter, you might be interested
+  in a series of articles, written by the Flutter Material team,
+  after they modified the [Flutter Gallery][] app to make it more
+  performant on the web:
+
+  * [Optimizing performance in Flutter web apps with tree
+    shaking and deferred loading][web-perf-1]
+  * [Improving perceived performance with image placeholders,
+    precaching, and disabled navigation transitions][web-perf-2]
+  * [Building performant Flutter widgets][web-perf-3]
+{{site.alert.end}}
+
+[Flutter Gallery]: {{site.gallery}}
+[web-perf-1]: {{site.flutter-medium}}/optimizing-performance-in-flutter-web-apps-with-tree-shaking-and-deferred-loading-535fbe3cd674
+[web-perf-2]: {{site.flutter-medium}}/improving-perceived-performance-with-image-placeholders-precaching-and-disabled-navigation-6b3601087a2b
+[web-perf-3]: {{site.flutter-medium}}/building-performant-flutter-widgets-3b2558aa08fa
+
+How do you design a Flutter app to most efficiently
+render your scenes? In particular, how do you ensure
+that the painting code generated by the
+framework is as efficient as possible?
+Some rendering and layout operations are known
+to be slow, but can't always be avoided.
+They should be used thoughtfully,
+following the guidance below.
+
+# Minimize expensive operations
+
+Some operations are more expensive than others,
+meaning that they consume more resources.
+Obviously, you want to only use these operations
+when necessary. How you design and implement your
+app's UI can have a big impact on how efficiently it runs.
+
+## Control build() cost
+
+Here are some things to keep in mind when designing your UI:
+
+* Avoid repetitive and costly work in `build()` methods
+  since `build()` can be invoked frequently when
+  ancestor widgets rebuild.
+* Avoid overly large single widgets with a large `build()` function.
+  Split them into different widgets based on encapsulation
+  but also on how they change:
+  * When `setState()` is called on a `State` object,
+    all descendent widgets rebuild. Therefore,
+    localize the `setState()` call to the part of
+    the subtree whose UI actually needs to change.
+    Avoid calling `setState()` high up in the tree
+    if the change is contained to a small part of the tree.
+  * The traversal to rebuild all descendents stops when the
+    same instance of the child widget as the previous
+    frame is re-encountered. This technique is heavily
+    used inside the framework for optimizing
+    animations where the animation doesn't affect the child subtree.
+    See the [`TransitionBuilder`][] pattern and
+    the [source code for `SlideTransition`][],
+    which uses this principle to avoid rebuilding its
+    descendents when animating.
+
+Use `const` constructors on widgets as much as possible,
```

Comment:
cross-linking opportunity: If apps have the recommended lints from package:flutter_lints enabled, they will automatically be reminded to use `const` where possible. If an app isn't using the recommended set of lints, they may want to consider upgrading to it as explained in https://docs.flutter.dev/release/breaking-changes/flutter-lints-package#migration-guide.